torch/optim/optimizer.py

```diff
@@ -403,6 +463,50 @@ def _patch_step_function(self) -> None:
             self.__class__.step = self.profile_hook_step(self.__class__.step)  # type: ignore[assignment]
             self.__class__.step.hooked = True  # type: ignore[attr-defined]
 
+    def _set_up_load_state_dict_pre_hooks_and_handle(self) -> None:
+        self._optimizer_load_state_dict_pre_hooks_legit: 'OrderedDict[int, Callable[["Optimizer", StateDict], Optional[StateDict]]]' = OrderedDict()  # noqa: B950
+        handle = hooks.RemovableHandle(self._optimizer_load_state_dict_pre_hooks_legit)
+        self._load_state_dict_cast_handle_legit: RemovableHandle = handle
+        self._optimizer_load_state_dict_pre_hooks_legit[handle.id] = Optimizer._cast_state_to_match_params_hook
+
+    @property
+    def _optimizer_load_state_dict_pre_hooks(
+            self) -> 'OrderedDict[int, Callable[["Optimizer", StateDict], Optional[StateDict]]]':
+        if (hasattr(self, "_optimizer_load_state_dict_pre_hooks_legit") and
+                self._optimizer_load_state_dict_pre_hooks_legit is not None):
+            return self._optimizer_load_state_dict_pre_hooks_legit
+        # otherwise, set up the hooks and handle
+        if (hasattr(self, "_load_state_dict_cast_handle_legit") and
+                self._load_state_dict_cast_handle_legit is not None):
+            raise RuntimeError("the load_state_dict pre hooks should always be init'd "
```

Comment:
Is that going to play well with serialization? Are we sure that there is no way we can have one but not the other? (both when users do the right thing of using state_dict and the wrong thing of serializing the whole Optimizer)